lucene/core/src/java/org/apache/lucene/index/IndexWriter.java

```diff
@@ -3252,6 +3317,53 @@ private long prepareCommitInternal() throws IOException {
       } finally {
         maybeCloseOnTragicEvent();
       }
+
+      if (mergeAwaitLatchRef != null && mergeAwaitLatchRef.get() != null) {
+        CountDownLatch mergeAwaitLatch = mergeAwaitLatchRef.get();
+        // If we found and registered any merges above, within the flushLock, then we want to ensure that they
+        // complete execution. Note that since we released the lock, other merges may have been scheduled. We will
+        // block until  the merges that we registered complete. As they complete, they will update toCommit to
+        // replace merged segments with the result of each merge.
+        mergeScheduler.merge(this, MergeTrigger.FULL_FLUSH, true);
+        long mergeWaitStart = System.nanoTime();
+        int abandonedCount = 0;
+        long waitTimeMillis = (long) (config.getMaxCommitMergeWaitSeconds() * 1000.0);
+        try {
+          config.getIndexWriterEvents().beginMergeOnCommit();
+          if (mergeAwaitLatch.await(waitTimeMillis, TimeUnit.MILLISECONDS) == false) {
+            synchronized (this) {
+              // Need to do this in a synchronized block, to make sure none of our commit merges are currently
+              // executing mergeFinished (since mergeFinished itself is called from with the IndexWriter lock).
+              // After we clear the value from mergeAwaitLatchRef, the merges we schedule will still execute as
+              // usual, but when they finish, they won't attempt to update toCommit or modify segment reference
+              // counts.
```

Comment:
Typically why would it happen that we abandon some merges? timeout? Would there be any sense in trying to interrupt them?

Reply:
Yeah -- maybe we only want to wait a short time for these merges to complete. If a merge doesn't complete, we should move forward with the commit. Since these merges will still update the "real" `SegmentInfos`, any progress they've made is arguably still useful. That's why I didn't want to interrupt them.